clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -188,21 +189,45 @@
     private static final String TOPIC_PARTITIONS_KEY_NAME = "previous_assignment";
     private static final String TOPIC_KEY_NAME = "topic";
     private static final String PARTITIONS_KEY_NAME = "partitions";
+    private static final String GENERATION_KEY_NAME = "generation";
     private static final Schema TOPIC_ASSIGNMENT = new Schema(
             new Field(TOPIC_KEY_NAME, Type.STRING),
             new Field(PARTITIONS_KEY_NAME, new ArrayOf(Type.INT32)));
     private static final Schema STICKY_ASSIGNOR_USER_DATA = new Schema(
-            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT)));
+            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT)),
+            new Field(GENERATION_KEY_NAME, Type.INT32));
```

Comment:
Have you thought about compatibility with the old schema? I am wondering if it would be better to somehow use the generation from the consumer group so that we don't have to change this schema.

Reply:
@hachikuji, thanks for the feedback, I had thought about it, but noticed it needed some updates. The new patch should address it.

Regarding using the existing consumer group generation, I think it may not worth the trouble and added complexity. We don't have consumer group generation on the client side (as far as I can tell), and protocol changes will be required. Plus, supporting the old schema seems to be simple enough. What do you think?

Reply:
Have you verified that this change is forward compatible as well? Older versions of the sticky assignor need to be able to work with the updated version.

The client definitely has access to the group's generation. The question is whether and how to expose it to the assigner.

Reply:
I tested this with a mix of old and new clients (using old and new versions of sticky assignor), and things seem to be running smoothly. The deserialization of new versions only looks at the partition array and ignores the generation marker.

Sorry, you're right. Generation seems to be available in the coordinator. But exposing it to just the sticky assignor could add complexity (other assignors don't need it). I still think the existing solution addresses the raised concern. But if you feel that using the consumer group generation is the way to go I can explore that approach. Thanks!

Reply:
@steven-aerts yes, that should still work if we decide to take that path. The leader always uses the latest generation and sends that along with the assigned partitions to each consumer in the group. Then later if a consumer drops for a brief period they would have a stale assignment (with an old generation). The leader knows to ignore those partition assignments in the case of a conflict. Would that make sense?

Reply:
It certainly makes sense, however, I do not see much added benefit compared to the more or less arbitrary generation counter you implemented in this PR.
If it was already exposed, you could use it, but I would not expose it just for this use case.

Reply:
Frankly I think you need a KIP for this change either way since you are changing the schema for the sticky assigner user data. Anything exposed on the wire is a public API. I'm actually a little skeptical that this solution can even work if you are using a local generation counter, but I have not reviewed the implementation in detail, so I could be wrong. 

The advantage of using the generation from the group is that it is guaranteed to increase monotonically after every rebalance. Exposing it in `PartitionAssignor.Assignment` seems reasonable at a glance and may be useful for other stateful assigners.

Reply:
I can write a short KIP. I have tested this in some common use cases and have not noticed an issue. I'll let you review this implementation, and I'll look at how your suggestion would work from an implementation point of view. Thanks!

Reply:
[KIP-341](https://cwiki.apache.org/confluence/display/KAFKA/KIP-341%3A+Update+Sticky+Assignor%27s+User+Data+Protocol) is created for this PR.